File: cli/npm/goleo/runtime/menu.go
package runtime

import (
	"errors"
	"fmt"
)

// MenuRole is a standard menu action wired to the platform's native handler
// (e.g. the responder chain on macOS), so the item works without a Go callback.
// Roles are what make Cmd+C/V/X/A/Z etc. work in the webview on macOS.
type MenuRole string

const (
	RoleNone      MenuRole = ""
	RoleQuit      MenuRole = "quit"
	RoleUndo      MenuRole = "undo"
	RoleRedo      MenuRole = "redo"
	RoleCut       MenuRole = "cut"
	RoleCopy      MenuRole = "copy"
	RolePaste     MenuRole = "paste"
	RoleSelectAll MenuRole = "selectAll"
	RoleMinimize  MenuRole = "minimize"
	RoleClose     MenuRole = "close"
)

// MenuItem is one entry in a native menu. A top-level MenuItem (with a Submenu)
// is a menu in the menu bar; nested items are its entries. Exactly one of Role /
// OnClick / Submenu / Separator is meaningful per item (Role wins over OnClick).
type MenuItem struct {
	Label       string
	Role        MenuRole
	Accelerator string // e.g. "cmd+q", "cmd+shift+z" (cmd/ctrl/alt/shift + key)
	OnClick     func()
	Submenu     []MenuItem
	Separator   bool
}

// SetMenu installs the application menu bar. Native on macOS; returns an
// errors.ErrUnsupported-wrapped error on Windows/Linux/mobile (no native menu
// bar yet — use an in-page HTML menu there). Safe to call after Run has started
// or from Config.Menu at startup.
func (a *App) SetMenu(menu []MenuItem) error {
	if !MenuSupported() {
		return fmt.Errorf("goleo: native menu: %w", errors.ErrUnsupported)
	}
	return a.setNativeMenu(menu)
}

// MenuSupported reports whether this platform has a native menu-bar backend
// (macOS today). Query it (or goleo:capabilities) before offering menu UI.
func MenuSupported() bool { return platformMenu }

// StandardMenu returns a conventional macOS menu bar — an App menu (Quit) and an
// Edit menu (undo/redo/cut/copy/paste/select-all) — so webview keyboard
// shortcuts work. Installed automatically on macOS when Config.Menu is empty;
// also a handy base to extend for custom menus.
func StandardMenu(appName string) []MenuItem {
	if appName == "" {
		appName = "App"
	}
	return []MenuItem{
		{Label: appName, Submenu: []MenuItem{
			{Label: "Quit " + appName, Role: RoleQuit, Accelerator: "cmd+q"},
		}},
		{Label: "Edit", Submenu: []MenuItem{
			{Label: "Undo", Role: RoleUndo, Accelerator: "cmd+z"},
			{Label: "Redo", Role: RoleRedo, Accelerator: "cmd+shift+z"},
			{Separator: true},
			{Label: "Cut", Role: RoleCut, Accelerator: "cmd+x"},
			{Label: "Copy", Role: RoleCopy, Accelerator: "cmd+c"},
			{Label: "Paste", Role: RolePaste, Accelerator: "cmd+v"},
			{Label: "Select All", Role: RoleSelectAll, Accelerator: "cmd+a"},
		}},
	}
}

File: cli/npm/goleo/bridge/src/menu.ts
import { getBridge } from './bridge'

/** One entry in a native menu (see App.SetMenu / Config.Menu on the Go side). */
export interface MenuItemSpec {
  /** Stable id; a leaf item with an id fires a `menu:<id>` event when clicked. */
  id?: string
  label?: string
  /**
   * Standard action wired to the OS (so Cmd/Ctrl shortcuts work): one of
   * quit, copy, paste, cut, selectAll, undo, redo, minimize, close.
   */
  role?: string
  /** e.g. "cmd+q", "cmd+shift+z" (macOS accelerators). */
  accelerator?: string
  separator?: boolean
  submenu?: MenuItemSpec[]
}

/**
 * Install the application menu bar natively. Resolves on all platforms with a
 * native menu bar (macOS/Windows/Linux); rejects with an ErrUnsupported-style
 * error elsewhere (PWA/mobile) — catch it and render an in-page menu instead.
 *
 * Handle clicks with onMenu(id, cb) (or bridge.on(`menu:${id}`, cb)).
 */
export async function setMenu(items: MenuItemSpec[]): Promise<void> {
  await getBridge().invoke('goleo:setMenu', { items })
}

/** Subscribe to clicks of the menu item with the given id. Returns unsubscribe. */
export function onMenu(id: string, cb: () => void): () => void {
  return getBridge().on(`menu:${id}`, () => cb())
}

/** Whether the running platform has a native menu bar (from goleo:capabilities). */
export async function menuSupported(): Promise<boolean> {
  const caps = await getBridge().invoke<Record<string, boolean>>('goleo:capabilities')
  return !!caps?.menu
}
